test/rulesets/Base/mapreduce.jl

```diff
@@ -38,4 +38,41 @@
             end
         end
     end  # sum abs2
+
+    @testset "prod" begin
+        @testset "Array{$T}" for T in [Float64, ComplexF64]
+            @testset "size = $sz, dims = $dims" for (sz, dims) in [
+                ((12,), :), ((12,), 1),
+                ((3,4), 1), ((3,4), 2), ((3,4), :), ((3,4), [1,2]),
+                ((3,4,1), 1), ((3,2,2), 3), ((3,2,2), 2:3),
+                ]
+                x, xdot, xbar = randn(T, sz), randn(T, sz), randn(T, sz)
+                # frule_test(prod, (x, xdot); fkwargs=(dims=dims,))
+                rrule_test(prod, prod(x; dims=dims), (x, xbar); fkwargs=(dims=dims,))
+
+                x[1] = 0
+                rrule_test(prod, prod(x; dims=dims), (x, xbar); fkwargs=(dims=dims,))
+
+                x[5] = 0
+                rrule_test(prod, prod(x; dims=dims), (x, xbar); fkwargs=(dims=dims,))
+
+                x[3] = x[7] = 0  # two zeros along some slice, for any dims
+                rrule_test(prod, prod(x; dims=dims), (x, xbar); fkwargs=(dims=dims,))
+
+                if ndims(x) == 3
+                    xp = PermutedDimsArray(x, (3,2,1))  # not a StridedArray
+                    xpdot, xpbar = permutedims(xdot, (3,2,1)), permutedims(xbar, (3,2,1))
+                    # frule_test(prod, (xp, xpdot); fkwargs=dims)
+                    rrule_test(prod, prod(xp; dims=dims), (xp, xpbar); fkwargs=(dims=dims,))
+                end
+            end
+        end
+        @testset "Array{Float32}" begin
+            v = [1f-5, 1f-10, 1f-15, 1f-20]
+            @test prod(v) == 0
+            vbar = randn(Float32, 4)
+            @test unthunk(rrule(prod, v)[2](1f0)[2]) == zeros(4)
+            rrule_test(prod, 0f0, (v, vbar))
```

Comment:
Does this actually test anything? Will this fail for non-zero `ȳ`?

Reply:
I guess this is just one more finite-diff test case, which happens to be in a weird corner. If something was broken perhaps you'd get NaN here.

The same vector of Float64 does not give zero, so I guess this tests a bit that you don't accidentally promote:
```
julia> v = [1f-5, 1f-10, 1f-15, 1f-20];

julia> unthunk(rrule(prod, v)[2](1f0)[2])
4-element Vector{Float32}:
 0.0
 0.0
 0.0
 0.0

julia> v64 = Float64[1f-5, 1f-10, 1f-15, 1f-20];

julia> unthunk(rrule(prod, v64)[2](1.0)[2])
4-element Vector{Float64}:
 9.999999852444477e-46
 9.99999946630892e-41
 9.999999563548297e-36
 9.999999917168005e-31
 ```

Reply:
Just to be clear, I am just proposing changing the `rrule_test` to something like:
```suggestion
            rrule_test(prod, rand(Float32), (v, vbar))
```
That should be fine, right?

Reply:
I thought `rrule_test` checked that the forward pass matches this 2nd argument, and thus this would fail?

Reply:
No, I believe the 2nd argument is the one passed to the pullback

Reply:
No, it seems it should be random.

Reply:
In which case I should change the others...

Reply:
Ah, yes, I missed those when reviewing. It still works, since it's the same shape as the output, but it's usually better to set those to a random value independently.

Reply:
Would be nice if rrule_test would just infer these, and all the perturbations, at least when they are just arrays.

Reply:
Yeah, might make sense to open an issue on ChainRulesTestUtils, if there isn't already. I think there are cases where you want to specify it manually to avoid some singularities though.